### PR TITLE
[Contrib] Add Authentication payload type to IKEv2

### PR DESCRIPTION
**Pull request by gpotter2**
Fixed version of https://github.com/secdev/scapy/pull/291
This version of the patch:
- was rebased & squashed
- has some small basic tests.
- was re-indented (tabs/spaces...)

Note: i have not reviewed every thing, some things might be ugly

### CODE REVIEW

File: scapy/contrib/ikev2.py [hunk 1]
```diff
@@ -313,6 +492,53 @@ class IKEv2_payload_VendorID(IKEv2_class):
         StrLenField("vendorID","",length_from=lambda x:x.length-4),
         ]
 
+class TrafficSelector(IKEv2_class):
```
Comment (guedou) [inline scapy/contrib/ikev2.py]: This is ugly.

Please use a dedicated class per traffic selector, and implement `guess_payload_class()` to do the dispatching.

File: scapy/contrib/ikev2.py [hunk 2]
```diff
@@ -397,8 +623,6 @@ class IKEv2_payload_IDr(IKEv2_class):
         StrLenField("load","",length_from=lambda x:x.length-8),
         ]
 
-
-
 class IKEv2_payload_Encrypted(IKEv2_class):
```
Comment (guedou) [inline scapy/contrib/ikev2.py]: Same comment here.

Comment (gpotter2) [inline scapy/contrib/ikev2.py]: Done

**Comment by guedou** [comment]
Looks good to me.

@rjarry what's your opinion ?

File: scapy/contrib/ikev2.py [hunk 3]
```diff
@@ -99,6 +101,19 @@
                                                  "ESN":   1,  }, 0),
                          }
 
+IKEv2AuthenticationTypes = {
+  0 : "Reserved",
+  1 : "RSA Digital Signature",
```
Comment (rjarry) [inline scapy/contrib/ikev2.py]: Indenting with 2 spaces is quite non-standard, but the file was already like that...

Maybe we should fix the indentation in a separate commit **before** this commit series.

File: scapy/contrib/ikev2.py [hunk 4]
```diff
@@ -432,13 +697,46 @@ class IKEv2_payload_CERTREQ(IKEv2_class):
...
+
+class IKEv2_payload_CERT_STR(IKEv2_payload_CERT):
+    name = "IKEv2 Certificate"
+    fields_desc = [
+        ByteEnumField("next_payload",None,IKEv2_payload_type),
+        ByteField("res",0),
+        FieldLenField("length",None,"cert_data","H",adjust=lambda pkt,x: x+5),
         ByteEnumField("cert_type",0,IKEv2CertificateEncodings),
-        StrLenField("cert_data","",length_from=lambda x:x.length-5),
+        StrLenField("cert_data", "", length_from=lambda x:x.length-5),
```
Comment (rjarry) [inline scapy/contrib/ikev2.py]: If you want to fix missing spaces after commas, you need to do it in a separate commit and also, fix the whole file.

File: scapy/contrib/ikev2.py [hunk 5]
```diff
@@ -456,12 +754,7 @@ class IKEv2_payload_CERT(IKEv2_class):
...
+    """Send a IKEv2 SA to an IP and wait for answers."""
     return sr(IP(dst=ip)/UDP()/IKEv2(init_SPI=RandString(8),
-                                      exch_type=34)/IKEv2_payload_SA(prop=IKEv2_payload_Proposal()))
-
-# conf.debug_dissector = 1
-
-if __name__ == "__main__":
-    from scapy.main import interact
-    interact(mydict=globals(), mybanner="IKEv2 alpha-level protocol implementation")
+                                      exch_type=34)/IKEv2_payload_SA(prop=IKEv2_payload_Proposal()), **kwargs)
```
Comment (rjarry) [inline scapy/contrib/ikev2.py]: That's not aligned to anything is it ?

File: scapy/contrib/ikev2.py [hunk 6]
```diff
@@ -435,16 +435,16 @@ def post_build(self, p, pay):
...
-
-	def i2h(self, pkt, x):
-		return IntField.i2h(self, pkt, x & 0xFFFF)
-		
-	def h2i(self, pkt, x):
-		return IntField.h2i(self, pkt, x if x !=None else 0 | 0x800E0000)
-		
+        # We only support the fixed-length Key Length attribute (the only one currently defined)
+        def __init__(self, name):
+                IntField.__init__(self, name, 0x800E0000)
```
Comment (rjarry) [inline scapy/contrib/ikev2.py]: If you replace tabs by spaces, use 1 tab -> 4 spaces, not 8

File: scapy/contrib/ikev2.py (hunk 5 above)
Comment (gpotter2) [inline scapy/contrib/ikev2.py]: To the line 759: https://github.com/secdev/scapy/pull/661/files#diff-487e064743c1edea916f1e38ec0175fbR759

File: scapy/contrib/ikev2.py (hunk 6 above)
Comment (gpotter2) [inline scapy/contrib/ikev2.py]: It's compulsory to replace tabs by space, if we want to support python 3.
Ok though.

File: scapy/contrib/ikev2.py (hunk 3 above)
Comment (gpotter2) [inline scapy/contrib/ikev2.py]: Sure.
When I fix an old patch, i useally leave the old work in a single piece. I'll add the commit just after that.
